84-sarcastic-fringehead/technical_specification.md

```diff
@@ -0,0 +1,478 @@
+- [ ] Get the updated `ghga-event-schemas` version
+- [ ] Adapt code for schema updates
+- [ ] Write migrations
+- [ ] Move migrations to own CLI command so they can be run as an init container
+  - [ ] Work with DevOps to get this configured in k8s
+
+## Diagrams:
+
+### Service Diagrams:
+#### Service Map
+![Service Map](./images/service_map.png)
+
+#### S3 Bucket Access Permissions
+![S3 Bucket Access Permissions](./images/bucket_access.png)
+
+#### Normal Upload Sequence Diagram (Beginning as the Upload Completes)
+
+```mermaid
+sequenceDiagram
+    box rgb(0, 150, 210, 0.5) Topics
+        participant FileUploads
+        participant FileUploadReports
+        participant ResearchDataUploadBoxes
+        participant FileInternallyRegistered
+    end
+    box rgb(255, 255, 200, .5) Services
+        participant Connector
+        participant UCS
+        participant UOS
+        participant EKSS
+        participant FIS
+        participant DHFS
+        participant DINS
+        participant IFRS
+    end
+    box rgb(200, 75, 35, 0.5) S3 Buckets
+        participant inbox
+        participant interrogation
+        participant archive
+    end
+    
+    Connector->>UCS: Complete Upload
+    UCS->>FileUploads: UPSERT: FileUpload(completed: True)
+    FileUploads->>FIS: UPSERT: FileUpload(completed: True)
+    FIS->>FIS: Create FileUploadReport with initial data
+    DHFS->>FIS: GET (polling)
+    FIS-->>DHFS: 200: list[FileUpload]
+    note right of DHFS: The following is<br>performed for<br>each file
+    DHFS->>inbox: Fetch first file chunk to get envelope
+    DHFS->>DHFS: Decrypt envelope with private key
+    DHFS->>DHFS: Generate new file secret
+    DHFS->>FIS: Submit new file secret (encrypted)
+    FIS->>EKSS: Submit secret
+    EKSS-->>FIS: Secret ID
+    DHFS->>interrogation: Initiate multipart upload
+    DHFS->>DHFS: Generate download URL
+    rect rgb(30, 30, 30, .4)
+        loop For each file chunk
+            inbox->>DHFS: Stream file chunk
+            DHFS->>DHFS: Decrypt with old secret
+            DHFS->>DHFS: Re-encrypt with new secret
+            DHFS->>DHFS: Update checksums
+            DHFS->>interrogation: Upload file chunk
+        end
+    end
+    DHFS->>DHFS: Compare checksums
+    DHFS->>interrogation: Complete multipart upload
+    DHFS->>FIS: POST InspectionResults
+    FIS->>FileUploadReports: Publish: FileUploadReport
+    FileUploadReports->>UCS: Consume: FileUploadReport
+    FileUploadReports->>IFRS: Consume: FileUploadReport
+    UCS->>inbox: Delete File
+    UCS->>FileUploads: UPSERT: FileUpload(state=ARCHIVED)
+    Note right of UCS: UCS considers items as<br> archived once they're <br> uploaded to the <br> interrogation bucket.
+    UOS->>ResearchDataUploadBoxes: UPSERT: ResearchDataUploadBox(state=CLOSED)
+    ResearchDataUploadBoxes->>IFRS: UPSERT: ResearchDataUploadBox(state=CLOSED)
+    IFRS->>interrogation: Copy File from interrogation to archive bucket
```

Comment:
Maybe note that this should happen for all gathered files that are inside the box.